lib/core/tools/line_tool/line_tool.dart

```diff
@@ -0,0 +1,302 @@
+// Dart imports:
+import 'dart:ui';
+
+// Flutter imports:
+import 'package:flutter/material.dart';
+
+// Package imports:
+import 'package:equatable/equatable.dart';
+
+// Project imports:
+import 'package:paintroid/core/commands/command_implementation/graphic/line_command.dart';
+import 'package:paintroid/core/commands/graphic_factory/graphic_factory.dart';
+import 'package:paintroid/core/commands/path_with_action_history.dart';
+import 'package:paintroid/core/tools/line_tool/line_tool_vertex.dart';
+import 'package:paintroid/core/tools/line_tool/vertex_stack.dart';
+import 'package:paintroid/core/tools/tool.dart';
+
+class LineTool extends Tool with EquatableMixin {
+  LineTool({
+    required super.paint,
+    required super.commandFactory,
+    required super.commandManager,
+    required this.graphicFactory,
+    required this.drawingSurfaceSize,
+    required super.type,
+  });
+
+  final GraphicFactory graphicFactory;
+  final Size drawingSurfaceSize;
+
+  VertexStack vertexStack = VertexStack();
+
+  Vertex? movingVertex;
+  Vertex? predecessorVertex;
+  Vertex? successorVertex;
+
+  LineCommand? ingoingGhostPathCommand;
+  LineCommand? outgoingGhostPathCommand;
+
+  bool addNewPath = false;
+
+  @override
+  List<Object?> get props => [commandManager, commandFactory, graphicFactory];
+
+  @override
+  void onDown(Offset point) {
+    if (vertexWasClicked(point)) {
+      return;
+    }
+
+    if (vertexStack.isEmpty) {
+      _createSourceAndDestinationCommandAndVertices(point);
+      return;
+    }
+
+    if (addNewPath) {
+      _createDestinationCommandAndVertex(point);
+      addNewPath = false;
+      return;
+    }
+  }
+
+  @override
+  void onDrag(Offset point) {
+    _setGhostPaths(point);
+  }
+
+  @override
+  void onUp(Offset point) {
+    _updateMovingVertices(point);
+  }
+
+  @override
+  void onCancel() {
+    reset();
+  }
+
+  void _setGhostPaths(Offset point) {
+    if (movingVertex == null) return;
+
+    point = getMovingVertexCenter(point);
+    movingVertex?.updateVertexCenter(point);
```

Comment:
Should it not be ! instead of ? ?

Reply:
! would guarantee that moving vertex is not null. I see you logic, becuase there is a nullcheck before using the variable, but dart treats it conservatively due to its global nature and i think its the safest if do the same and use ?.